src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -71,6 +74,192 @@ static void Main()
+            var model = comp.GetSemanticModel(tree);
+            var declarators = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First();
+            var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarators);
+            Assert.Equal("T", symbol.Type.ToTestDisplayString());
+            Assert.Equal(false, symbol.Type.IsNullable);
+        }
+
+        [Fact]
+        public void Default_TClass()
+        {
+            var source =
+@"class C
+{
+    static void F<T>() where T : class
+    {
+        var s = default(T);
+        s.ToString();
+        var t = default(T?);
+        t.ToString();
+    }
+}";
+
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,17): warning CS8599: Creating a null value of a non-nullable type.
+                //         var s = default(T);
+                Diagnostic(ErrorCode.WRN_DefaultOfNonNullableType, "default(T)").WithLocation(5, 17),
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         t.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t").WithLocation(8, 9));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var declarators = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First();
+            var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarators);
+            Assert.Equal("T", symbol.Type.ToTestDisplayString());
+            Assert.Equal(false, symbol.Type.IsNullable);
+        }
+
+        [Fact]
+        public void DefaultInferred_NonNullable()
+        {
+            var source =
+@"class C
+{
+    static void Main()
+    {
+        string s = default;
+        s.ToString();
+        int i = default;
+        i.ToString();
+    }
+}";
+
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,20): warning CS8600: Cannot convert null to non-nullable reference.
+                //         string s = default;
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "default").WithLocation(5, 20));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var declarators = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First();
+            var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarators);
+            Assert.Equal("System.String", symbol.Type.ToTestDisplayString());
+            Assert.Equal(false, symbol.Type.IsNullable);
+        }
+
+        [Fact]
+        public void DefaultInferred_Nullable()
+        {
+            var source =
+@"class C
+{
+    static void Main()
+    {
+        string? s = default;
```

Comment:
>string? s = default; [](start = 8, length = 20)

I'm curious what happens with `string? s = default!;`. Where did LDM leave this off? Does ! affect the flow analysis, or just the warning at that spot?